draft-irtf-cfrg-fiat-shamir.md

```diff
@@ -107,59 +113,128 @@ This will be expanded in future versions of this specification.
 We describe how to construct non-interactive proofs for sigma protocols.
 The Fiat-Shamir transformation is parametrized by:
 
+- a `SigmaProtocol`, which specifies an interactive 3-message protocol as defined in {{Section 2 of !SIGMA=I-D.draft-irtf-cfrg-sigma-protocols-00}};
 - a `Codec`, which specifies how to absorb prover messages and how to squeeze verifier challenges;
-- a `SigmaProtocol`, which specifies an interactive 3-message protocol.
+- a `DuplexSpongeInterface`, which specifies a hash function for computing challenges.
 
-Upon initialization, the protocol receives as input an `iv` of 32-bytes which uniquely identifies the protocol and the session being proven and (optionally) pre-processes some information about the protocol using the instance.
+Upon initialization, the protocol receives as input:
+- `session_id`, which identifies the session being proven
+- `instance`, the sigma protocol instance for proving or verifying
 
     class NISigmaProtocol:
-        Protocol: SigmaProtocol
-        Codec: Codec
+        Protocol: SigmaProtocol = None
+        Codec: Codec = None
+        Hash: DuplexSpongeInterface = None
 
-        def init(self, iv: bytes, instance):
+        def __init__(self, session_id, instance):
             self.hash_state = self.Codec(iv)
             self.ip = self.Protocol(instance)
 
-        def prove(self, witness, rng):
-            (prover_state, commitment) = self.ip.prover_commit(witness, rng)
-            challenge = self.hash_state.prover_message(commitment).verifier_challenge()
-            response = self.ip.prover_response(prover_state, challenge)
+        def _prove(self, witness, rng):
+            # Core proving logic that returns commitment, challenge, and response.
+            # The challenge is generated via the hash function.
+            (prover_state, commitment) = self.sigma_protocol.prover_commit(witness, rng)
+            self.codec.prover_message(self.hash_state, commitment)
+            challenge = self.codec.verifier_challenge(self.hash_state)
+            response = self.sigma_protocol.prover_response(prover_state, challenge)
+            return (commitment, challenge, response)
 
-            assert self.ip.verifier(commitment, challenge, response)
-            return self.ip.serialize_commitment(commitment) + self.ip.serialize_response(response)
+        def prove(self, witness, rng):
+            # Default proving method using challenge-response format.
+            (commitment, challenge, response) = self._prove(witness, rng)
+            assert self.sigma_protocol.verifier(commitment, challenge, response)
+            domain_sep = "NISigmaProtocolShortProof".encode('utf-8')
```

Comment:
It looks like you need to update the code snippets once more to match the reference implementation.